test-d/is-tuple.ts

```diff
@@ -0,0 +1,59 @@
+import {expectType} from 'tsd';
+import type {IsTuple} from '../index';
+
+// Tuples
+expectType<IsTuple<[]>>(true);
+expectType<IsTuple<[number]>>(true);
+expectType<IsTuple<[number, string]>>(true);
+expectType<IsTuple<[number, string, ...number[]]>>(true);
+expectType<IsTuple<[number?]>>(true);
+expectType<IsTuple<[number?, string?]>>(true);
+expectType<IsTuple<[number?, string?, ...number[]]>>(true);
+expectType<IsTuple<[never]>>(true);
+
+// Readonly tuples
+expectType<IsTuple<readonly []>>(true);
+expectType<IsTuple<readonly [number]>>(true);
+expectType<IsTuple<readonly [number, string, ...number[]]>>(true);
+expectType<IsTuple<readonly [number?, string?, ...number[]]>>(true);
+expectType<IsTuple<readonly [number?]>>(true);
+expectType<IsTuple<readonly [never]>>(true);
+
+// Non-tuples
+expectType<IsTuple<number[]>>(false);
+expectType<IsTuple<readonly number[]>>(false);
+expectType<IsTuple<[...number[]]>>(false);
```

Comment:
What's these expected results?
```ts
IsTuple<[...number[], string]
IsTuple<[...number[], number]
IsTuple<[number, string, ...number[]]
IsTuple<[number, ...number[]]
```

Reply:
@Emiyaaaaa Interesting!

The bottom two cases are tuples when `fixedLengthOnly` is `false`, and they are not otherwise. And there's already test cases for these.

The top two cases seem like they should be tuples, but the current implementation doesn't handle these. Let me find a fix!

Reply:
Wait, what's these expected results?
```
IsTuple<number[]>
IsTuple<[...number[]]
IsTuple<[...number[], number]
```

Reply:
> Wait, what's these expected results?
> 
> ```ts
> IsTuple<number[]>
> IsTuple<[...number[]]
> IsTuple<[...number[], number]
> ```


Maybe all non-fixed-length array sould not be tuple?

Reply:
The first two should always be `false`. 
The third one should be a tuple when `fixedLengthOnly` is `false`.

Reply:
Meybe all non-fixed-length array should not be tuple? The definition of 'tuple' is an immutable array.

Reply:
There's an option to control this behaviour. In TS, there might be cases where distinguishing between `number[]` and `[number, ...number[]]` makes sense.

Maybe we can the default behaviour though?

Reply:
> there might be cases where distinguishing between `number[]` and `[number, ...number[]]` would make sense.

But why use `IsTuple` to distinguishing between `number[]` and `[number, ...number[]]` and not something else? look like confusion.

Some related type: [StaticPartOfArray](https://github.com/sindresorhus/type-fest/blob/3d546271332116970e5facde69b4447d2e64b45f/source/internal/array.d.ts#L42), [VariablePartOfArray](https://github.com/sindresorhus/type-fest/blob/3d546271332116970e5facde69b4447d2e64b45f/source/internal/array.d.ts#L61C13-L61C32)

> Maybe we can update the default behaviour though?
I agree

Reply:
> But why use IsTuple to distinguishing between number[] and [number, ...number[]] and not something else? look like confusion.

Umm...I meant just for distinguishing, like either or cases. True that [StaticPartOfArray](https://github.com/sindresorhus/type-fest/blob/3d546271332116970e5facde69b4447d2e64b45f/source/internal/array.d.ts#L42), [VariablePartOfArray](https://github.com/sindresorhus/type-fest/blob/3d546271332116970e5facde69b4447d2e64b45f/source/internal/array.d.ts#L61C13-L61C32) can also be used, but IMO it doesn't hurt to give that option with `IsTuple` as well.

WDTY? @sindresorhus @Snowflyt

Reply:
@Emiyaaaaa 
> The top two cases seem like they should be tuples, but the current implementation doesn't handle these. Let me find a fix!

Pushed a fix for this in [985ee21](https://github.com/sindresorhus/type-fest/pull/1024/commits/985ee21e35d47ec2850d11c2723b319376a3ef04), not the most elegant fix, but it works. WDYT?

Reply:
great!

Reply:
> but IMO it doesn't hurt to give that option with `IsTuple` as well.

we can set default options to true first.

Reply:
Updated the default in [dd630e6](https://github.com/sindresorhus/type-fest/pull/1024/commits/dd630e60fa112389a9fd0a25e3af48af56898d49).